File: firstnlast.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if(len(nums) == 0):
            return [-1,-1]
        if (target < nums[0] or target > nums[len(nums) - 1]):
            return [-1,-1]
        first = self.bsleft(nums,target)
        if(first == -1):
            return [-1,-1]
        last = self.bsright(nums,target)
        return [first,last]

        
    def bsleft(self,nums,target):
        low = 0
        high = len(nums) - 1
        while(low<=high):
            mid = int(low + (high - low)/2)
            if(nums[mid]==target):
                if(mid == 0 or nums[mid-1]<nums[mid]):
                    return mid
                else:
                    high = mid -1
            elif(target < nums[mid]):
                high = mid -1
            else:
                low = mid + 1
        return -1
    
    def bsright(self,nums,target):
        low = 0
        high = len(nums) - 1
        while(low<=high):
            mid = int(low + (high - low)/2)
            if(nums[mid]==target):
                if(mid == len(nums) -1 or nums[mid+1]>nums[mid]):
                    return mid
                else:
                    low = mid + 1
            elif(target < nums[mid]):
                high = mid -1
            else:
                low = mid + 1
        return -1

File: firstNlast.java
class Solution {
    public int first(int[] nums, int target, int low, int high){
        // binary search for 1st index
        while(low<=high){
            int mid = low + (high-low)/2;
            if(nums[mid]==target){
                if(mid == low || nums[mid-1] < nums[mid]){
                    return mid;
                } else {
                    high = mid -1;
                }
                
            } else if(nums[mid]> target){
                high = mid -1;
            } else  if(nums[mid]<target){
                low = mid +1;
            }
        }
        return -1;
    }
    public int second(int[] nums, int target, int low, int high){
        // binary search for 1st index
        while(low<=high){
            int mid = low + (high-low)/2;
            if(nums[mid]==target){
                if(mid == high || nums[mid] < nums[mid+1]){
                    return mid;
                } else {
                    low = mid+1;
                }
                
            } else if(nums[mid]> target){
                high = mid -1;
            } else  if(nums[mid]<target){
                low = mid +1;
            }
        }
        return -1;
    }
    public int[] searchRange(int[] nums, int target) {
        // null case
        if(nums.length == 0){
            return new int[] {-1,-1};
        } 
        int n1 = first(nums, target, 0, nums.length-1);
        if(n1 == -1) return new int[] {-1,-1};
        int n2 = second(nums, target, 0, nums.length-1);
        return new int[] {n1,n2};
        
    }
}

// Time Complexity: O(logN)
// Space Complexity: O(1)
